### PR TITLE
add Psychic reader to llama_index

### PR DESCRIPTION
**Pull request by jasonwcfan**
Adding a new Reader to LlamaIndex.

Psychic: https://github.com/psychic-api/psychic
Psychic docs: https://docs.psychic.dev/introduction

Psychic is an open source tool that makes it easy to set up auth + data syncs with a variety of SaaS data sources, normalize data as standard HTML, and convert this data to LlamaIndex-compatible Documents.

Reason for adding it as a LlamaIndex Reader rather than a LlamaHub data loader:
* Psychic is an aggregator of data sources, not a single data source unlike other LlamaHub loaders.
* The `download_loader` pattern doesn't work well for some Psychic users who want to avoid having a dependency on a network call to github

### CODE REVIEW

**Comment by jasonwcfan** [comment]
Hi @jerryjliu ! As promised here's a first pass at a Psychic integration with LlamaIndex as a Reader. Let me know what you think, open to feedback.

**Comment by Disiok** [comment]
Hey @jasonwcfan , great to see this. Will take a quick initial look!

**Comment by jasonwcfan** [comment]
Thanks @Disiok !

File: llama_index/readers/psychic.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            if psychic_key is None:
+                raise ValueError(
+                    "Must specify `psychic_key` or set environment "
+                    "variable `PSYCHIC_SECRET_KEY`."
+                )
+
+        self.psychic = Psychic(secret_key=psychic_key)
+        self.ConnectorId = ConnectorId
+
+    def load_data(self, connector_id: str = None, account_id: str = None) -> List[Document]:
```
Comment (Disiok) [inline llama_index/readers/psychic.py:45]: To clarify. This reads "all data" from data sources configured in the Psychic dashboard? Curious is there away to configure which subset of data is read?

File: llama_index/readers/psychic.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        if connector_id not in self.ConnectorId.__members__.keys():
+            raise ValueError("Invalid connector ID.")
+        
+        # get all the documents in the database
+        docs = []
+        data = self.psychic.get_documents(self.ConnectorId[connector_id], account_id)
+        for resource in data:
+            text = resource.get("content")
+            doc_id = resource.get("uri")
+            docs.append(Document(text=text, doc_id=doc_id))
```
Comment (Disiok) [inline llama_index/readers/psychic.py:67]: curious: are there other metadata that's worthwhile to expose? You can inject it as `extra_info`

File: docs/examples/data_connectors/PsychicDemo.ipynb [hunk 3]
```diff
@@ -0,0 +1,98 @@
+{
```
Comment (Disiok) [inline docs/examples/data_connectors/PsychicDemo.ipynb:1]: very nit: would prefer if you keep the outputs in the notebook.

Comment (Disiok) [reply docs/examples/data_connectors/PsychicDemo.ipynb:1]: Could you also link this into our documentation via `docs/how_to/data_connectors.md` See bottom of the page for adding to the `Examples` toctree.

**Review by Disiok — APPROVED** [review]
Great! This looks straightforward enough.

Would appreciate if you can add a section to your docs for how to use with LlamaIndex (similar to https://docs.psychic.dev/langchain) 🙏

File: llama_index/readers/psychic.py (hunk 1 above)
Comment (jasonwcfan) [reply llama_index/readers/psychic.py:45]: It reads only the data associated with a certain account (a Psychic user's user) and only for a specific source, not all sources.

We also have metadata filters, but it's only partially implemented right now.

Comment (Disiok) [reply llama_index/readers/psychic.py:45]: Cool, thanks for clarifying!

File: docs/examples/data_connectors/PsychicDemo.ipynb (hunk 3 above)
Comment (jasonwcfan) [reply docs/examples/data_connectors/PsychicDemo.ipynb:1]: Will do on both counts! re: outputs that's just a matter of not using the logger and not using `display()`?

File: llama_index/readers/psychic.py (hunk 2 above)
Comment (jasonwcfan) [reply llama_index/readers/psychic.py:67]: Yes! Probably the `account_id` and `connector_id`, I'll add those. We're also adding support for metadata filtering soon too.

**Comment by jasonwcfan** [comment]
> Great! This looks straightforward enough.
> 
> Would appreciate if you can add a section to your docs for how to use with LlamaIndex (similar to https://docs.psychic.dev/langchain) 🙏

Of course! Made the changes suggested - I'll update our own docs as well tomorrow

**Comment by jerryjliu** [comment]
> > Great! This looks straightforward enough.
> > Would appreciate if you can add a section to your docs for how to use with LlamaIndex (similar to https://docs.psychic.dev/langchain) 🙏
> 
> Of course! Made the changes suggested - I'll update our own docs as well tomorrow

this is great! btw @jasonwcfan we're refactoring llama-hub to be a python package too, should be a lot easier for you to submit the reader there too once that lands

**Comment by jasonwcfan** [comment]
Hey @Disiok made some changes based on feedback + updated the `__init__.py` files. Let me know if this looks good to you

**Comment by Disiok** [comment]
Nice! Excited to get this in!

**Pull request merged**